SpiceSharp/Components/Waveforms/Pwl.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Collections.Generic;
+using SpiceSharp.IntegrationMethods;
+using SpiceSharp.Simulations;
+
+namespace SpiceSharp.Components.Waveforms
+{
+    public class Pwl : Waveform
+    {
+        public Pwl(double[] times, double[] voltages, double initialVoltage = 0.0)
+        {
+            if (times == null)
+            {
+                throw new NullReferenceException(nameof(times));
+            }
+
+            if (voltages == null)
+            {
+                throw new NullReferenceException(nameof(voltages));
+            }
+
+            if (times.Length != voltages.Length)
+            {
+                throw new InvalidOperationException("PWL - times array has different length than voltages array");
+            }
+
+            Points = CreatePoints(times, voltages, initialVoltage);
+            LineParameters = CreateLineParameters(Points);
+        }
+
+        public List<Point> Points { get; }
+
+        protected LineDefinition[] LineParameters { get; }
+
+        public override void Accept(TimeSimulation simulation)
+        {
+            if (simulation == null)
+            {
+                throw new ArgumentNullException(nameof(simulation));
+            }
+
+            double time = simulation.Method.Time;
+            if (time.Equals(0.0))
+                Value = Points[0].Y;
+
+            if (simulation.Method is IBreakpoints method)
+            {
+                var breaks = method.Breakpoints;
+                if (!method.Break)
+                    return;
+
+                for (int i = 0; i < Points.Count; i++)
```

Comment:
Every time a breakpoint is hit, you add all points again to the breakpoint list. Either add the breakpoints only once or only add the next breakpoint. The second option is preferred if possible.

Reply:
You can also calculate the slope and intercept here.

Reply:
@svenboulanger  For now I will go for creating all breakpoints only once. It has the advantage that waveform will have all the corners from pwl despite max step / step of transient.

Reply:
The break point system already manages this. A breakpoint will force the time analysis to stop at a predetermined time point.

Reply:
@svenboulanger OK. I will try to change it. Please take a look at latests commits.